rfcs/0007-synchronous-commit.md

```diff
@@ -0,0 +1,162 @@
+* `remote_apply`: Commits wait for the WAL to be written and flushed locally, then wait for standbys to fully apply the WAL, this ensures the data consistent between primary and standby.
+
+The article "Understanding synchronous_commit in PostgreSQL" includes a helpful diagram showing how the `on` and `off` settings work:
+
+![](./images/postgres-sync-commit.png)
+
+The key distinction between `on` and `off` is that `on` ensures the WAL is written and flushed to local storage before proceeding. This highlights a central theme throughout this RFC: synchronous commit and durability fundamentally revolve around WAL handling.
+
+Let's summarize the use cases and trade-offs for different `synchronous_commit` levels:
+
+1. For financial systems where data is highly sensitive and data loss is unacceptable, `on`, `remote_write`, or `remote_apply` should be used.
+2. For mission-critical systems that cannot tolerate data inconsistency between primary and standby servers after a failover, `remote_apply` should be used to guarantee data consistency across servers.
+3. For workloads like logging or stream processing where some data loss is acceptable and performance is paramount, `off` or `local` can be used to optimize throughput.
+
+### RocksDB
+
+Let's look at how RocksDB describes synchronous commit in their documentation:
+
+> #### Non-Sync Mode
+>
+> When WriteOptions.sync = false (the default), WAL writes are not synchronized to disk. Unless the operating system thinks it must flush the data (e.g. too many dirty pages), users don't need to wait for any I/O for write.
+>
+> Users who want to even reduce the CPU of latency introduced by writing to OS page cache, can choose Options.manual_wal_flush = true. With this option, WAL writes are not even flushed to the file system page cache, but kept in RocksDB. Users need to call DB::FlushWAL() to have buffered entries go to the file system.
+>
+> Users can call DB::SyncWAL() to force fsync WAL files. The function will not block writes being executed in other threads.
+>
+> In this mode, the WAL write is not crash safe.
+>
+> #### Sync Mode
+>
+> When WriteOptions.sync = true, the WAL file is fsync'ed before returning to the user.
+>
+> #### Group Commit
+>
+> As most other systems relying on logs, RocksDB supports group commit to improve WAL writing throughput, as well as write amplification. RocksDB's group commit is implemented in a naive way: when different threads are writing to the same DB at the same time, all outstanding writes that qualify to be combined will be combined together and write to WAL once, with one fsync. In this way, more writes can be completed by the same number of I/Os.
+>
+> Writes with different write options might disqualify themselves to be combined. The maximum group size is 1MB. RocksDB won't try to increase batch size by proactive delaying the writes.
+
```

Comment:
Just leaving a not here: as I understand it, RocksDB has 3 distinct write levels: WAL (manual_wal_flush=true, sync=false), page cache (manual_wal_flush=false, sync=false), on disk (manual_wal_flush=false, sync=true). SlateDB has no page cache, so we have only WAL and remote.

Reply:
oh, that's accurate, `manual_wal_flush` does not even write to the page cache. i've missed this point, let me fix the description here :pray: